### PR TITLE
Mark tasks as did work even when all they did was remove stale output

### LINKED ISSUE #10418: JavaCompile is UP-TO-DATE even if it had to remove stale classes
Looks like we are ignoring whether `StaleClassCleaner` did any work.

We could use this opportunity to simplify `StaleClassCleaner`'s API.

---
cc: @gradle/build-cache

### PR DESCRIPTION
**Pull request by lptr**
Fixes #10418

### CODE REVIEW

File: subprojects/platform-base/src/test/groovy/org/gradle/language/base/internal/tasks/StaleOutputCleanerTest.groovy [hunk 1]
```diff
@@ -22,106 +22,64 @@ import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
...
 
     def "deletes empty parent directories"() {
         def file1 = tmpDir.file('foo/bar/file1').createFile()
         tmpDir.file('foo/baz/file2').createFile()
-        def cleaner = new SimpleStaleClassCleaner(deleter, files(file1))
-        cleaner.addDirToClean(tmpDir.testDirectory)
 
-        when:
-        cleaner.execute()
+        expect:
```
Comment (wolfs) [inline subprojects/platform-base/src/test/groovy/org/gradle/language/base/internal/tasks/StaleOutputCleanerTest.groovy]: Nit: extra new line

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/tasks/AbstractLinkTask.java [hunk 2]
```diff
@@ -226,12 +220,14 @@ protected Deleter getDeleter() {
 
     @TaskAction
     public void link() {
-        SimpleStaleClassCleaner cleaner = new SimpleStaleClassCleaner(getDeleter(), getOutputs());
-        cleaner.addDirToClean(getDestinationDirectory().get().getAsFile());
-        cleaner.execute();
+        boolean cleanerOutputs = StaleOutputCleaner.cleanOutputs(
```
Comment (wolfs) [inline subprojects/platform-native/src/main/java/org/gradle/nativeplatform/tasks/AbstractLinkTask.java]: I suppose you meant `cleanedOutputs`.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CleaningJavaCompiler.java [hunk 3]
```diff
@@ -15,30 +15,53 @@
...
 
-public class CleaningJavaCompiler extends CleaningJavaCompilerSupport<JavaCompileSpec> implements org.gradle.language.base.internal.compile.Compiler<JavaCompileSpec> {
-    private final Compiler<JavaCompileSpec> compiler;
+import javax.annotation.Nullable;
+import java.io.File;
+
+/**
+ * Deletes stale classes before invoking the actual compiler.
+ */
+public class CleaningJavaCompiler<T extends JavaCompileSpec> implements Compiler<T> {
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CleaningJavaCompiler.java]: I thought the same, but it is actually about Java in that we only accept a `JavaCompileSpec`, not a `JvmCompileSpec`. I was toying with `CleaningJavaLikeCompiler` but decided against.

File: subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CleaningJavaCompiler.java [hunk 4]
```diff
@@ -15,30 +15,53 @@
...
         this.compiler = compiler;
         this.taskOutputs = taskOutputs;
         this.deleter = deleter;
     }
 
     @Override
-    public Compiler<JavaCompileSpec> getCompiler() {
-        return compiler;
+    public WorkResult execute(T spec) {
+
```
[the remark this answers is not in the document]
Comment (lptr) [reply subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/CleaningJavaCompiler.java]: Fixed.

File: subprojects/platform-base/src/test/groovy/org/gradle/language/base/internal/tasks/StaleOutputCleanerTest.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/platform-base/src/test/groovy/org/gradle/language/base/internal/tasks/StaleOutputCleanerTest.groovy]: Fixed.

File: subprojects/platform-native/src/main/java/org/gradle/nativeplatform/tasks/AbstractLinkTask.java (hunk 2 above)
Comment (lptr) [reply subprojects/platform-native/src/main/java/org/gradle/nativeplatform/tasks/AbstractLinkTask.java]: Fixed.

**Pull request merged**